new-log-viewer/src/contexts/UrlContextProvider.tsx

```diff
@@ -0,0 +1,209 @@
+import React, {
+    createContext,
+    useEffect,
+    useState,
+} from "react";
+import {
+    SEARCH_PARAM_NAME,
+    HASH_PARAM_NAME,
+    UrlSearchParams,
+    UrlHashParams,
+    UrlSearchParamUpdatesType,
+    UrlHashParamUpdatesType,
+    UrlParamsType
+} from "../typings/url";
+
+
+const UrlContext = createContext <UrlParamsType>({} as UrlParamsType);
+
+interface UrlContextProviderProps {
+    children: React.ReactNode
+}
+
+let searchParams = new URLSearchParams(window.location.search.substring(1));
+
+/**
+ * Computes updated URL search parameters based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the search parameters.
+ * If a key's value is `null`, the key will be removed from the search parameters.
+ * @return - The updated search string as a URLSearchParams object.
+ */
+const getUpdatedSearchParams = (updates: UrlSearchParamUpdatesType) => {
+    const newSearchParams = searchParams;
+    const {filePath} = updates;
+
+    for (const [key, value] of Object.entries(updates)) {
+        if ("filePath" === key) {
+            continue;
+        }
+        if (null === value) {
+            newSearchParams.delete(key);
+        } else {
+            newSearchParams.set(key, String(value));
+        }
+    }
+    if ("string" === typeof filePath) {
+        newSearchParams.set("filePath", filePath);
+    }
+
+    return newSearchParams;
+};
+
+/**
+ * Computes updated URL search parameters based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the hash parameters.
+ * If a key's value is `null`, the key will be removed from the hash parameters.
+ * @return - The updated search string as a URLSearchParams object.
+ */
+const getUpdatedHashParams = (updates: UrlHashParamUpdatesType) => {
+    const newHashParams = new URLSearchParams(window.location.hash.substring(1));
+    for (const [key, value] of Object.entries(updates)) {
+        if (null === value) {
+            newHashParams.delete(key);
+        } else {
+            newHashParams.set(key, String(value));
+        }
+    }
+
+    return newHashParams;
+};
+
+/**
+ * Updates search parameters in the current window's URL based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the search parameters.
+ * If a key's value is `null`, the key will be removed from the search parameters.
+ */
+const updateWindowSearchParams = (updates: UrlSearchParamUpdatesType) => {
+    const newUrl = new URL(window.location.href);
+    searchParams = getUpdatedSearchParams(updates);
```

Comment:
This makes sense, but we want to have `searchParams` as an object storing the parameters that have been parsed, right?